### PR TITLE
feat: reintroduce usernames

### PR DESCRIPTION
**Pull request by binarious**
Please test this locally.

**Warning:** We need to manually do a `db.users.dropIndex('username')`, because some collections had this field. Otherwise we'll get:
```
duplicate key error collection: test.users index: username_1 dup key
```

### CODE REVIEW

File: src/config/passport.conf.js [hunk 1]
```diff
@@ -8,6 +8,8 @@ import User from '../app/models/user.model.js';
 import md5 from 'md5';
 import randomstring from 'randomstring';
 
+const reservedUsernames = ['yuttadhammo', 'doug', 'martin', 'robin', 'sebastian', 'it'];
+
 export default (passport) => {
   // Define length boundaries for expected parameters
   let bounds = {
```
Comment (olastor) [inline src/config/passport.conf.js]: The minimum email length of 5 characters for email prevents someone with a shorter username to login.

File: src/config/passport.conf.js [hunk 2]
```diff
@@ -118,50 +124,83 @@ export default (passport) => {
...
+        }
+
+        // If the user's email address is not verified yet
+        if (!user.verified && (!user.role || user.role !== 'ROLE_ADMIN')) {
+          return done(null, false, { loginMessage: 'Please confirm your email address.'});
+        }
+
+        // Check if a username is set and if not request one
+        if (!user.username) {
+          if (!req.body.username || reservedUsernames.contains(req.body.username)) {
```
Comment (olastor) [inline src/config/passport.conf.js]: Throws a Type Error for me? I don't know a `Array.contains()` method, did you mean `reservedUsernames.indexOf(req.body.username) !== -1`

File: src/config/passport.conf.js [hunk 3]
```diff
@@ -74,18 +76,22 @@ export default (passport) => {
...
+        $or: [
+          { 'local.email': email },
+          { 'username': req.body.username }
+        ]
       }, (err, user) => {
         if (err) return done(err);
 
-        if (user) {
-          // Email already taken.
+        if (user || reservedUsernames.contains(req.body.username) ) {
```
Comment (olastor) [inline src/config/passport.conf.js]: same thing as below

File: src/config/passport.conf.js [hunk 4]
```diff
@@ -8,6 +8,8 @@ import User from '../app/models/user.model.js';
 import md5 from 'md5';
 import randomstring from 'randomstring';
 
+const reservedUsernames = ['yuttadhammo', 'doug', 'martin', 'robin', 'sebastian', 'it'];
```
Comment (olastor) [inline src/config/passport.conf.js]: good idea! the username `it` can't be used with the current validation (username must be longer than 3 chars), though.

**Review by olastor — CHANGES_REQUESTED** [review]
The server crashes (locally) if I try to sign up a new account and choose an existing name (not the username): `error: uncaughtException: insertDocument :: caused by :: 11000 E11000 duplicate key error index: test.users.$name_1  dup key: { : "Sebastian23" }`

That's some weird behaviour, can you reproduce it?

File: src/config/passport.conf.js (hunk 2 above)
Comment (binarious) [reply src/config/passport.conf.js]: Whoops, I meant [includes](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/includes). Will fix it.

**Comment by binarious** [comment]
@olastor Can you try it with a fresh database?

**Comment by olastor** [comment]
@binarious Ok worked now.

**Comment by binarious** [comment]
@olastor Things should be fixed.

**Comment by binarious** [comment]
@olastor Tests are still failing and I can't find the error at the moment. Could you take a look?

**Comment by olastor** [comment]
@binarious I'll try to solve it.

**Review by olastor — APPROVED** [review]

**Pull request merged**